sql/sql_regex.h

```diff
@@ -0,0 +1,132 @@
+/* Copyright (c) 2016, Percona and/or its affiliates. All rights reserved.
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; version 2 of the License.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA */
+
+#ifndef _regex_list_handler_h_
+#define _regex_list_handler_h_
+
+#include <string>
+#include <algorithm>
+
+#include "my_global.h"
+#include "my_pthread.h"
+#include "my_regex.h"
+
+/*
+  Wrapper class aroung my_regex that manages concurrent acess and the lifetime
+  of the compiled regex.
+*/
+class Regex
+{
+private:
+  std::string m_pattern;
+  my_regex_t m_expr;
+  bool m_compiled;
+
+  mutable mysql_rwlock_t m_rwlock;
+
+  Regex(const Regex& other) {}
+  Regex& operator=(const Regex& other)
+  {
+    return *this;
+  }
+
+public:
+  Regex() :
+    m_pattern(""),
+    m_compiled(false)
+  {
+    mysql_rwlock_init(0, &m_rwlock);
+  }
+
+  ~Regex()
+  {
+    mysql_rwlock_wrlock(&m_rwlock);
+    if (m_compiled)
+    {
+      my_regfree(&m_expr);
+      m_compiled = false;
+    }
+    mysql_rwlock_unlock(&m_rwlock);
+    mysql_rwlock_destroy(&m_rwlock);
+  }
+
+  bool compiled() const
+  {
+    return m_compiled;
+  }
+
+  const std::string& pattern() const
+  {
+    return m_pattern;
+  }
+
+  // Compile the pattern into an expression
+  bool compile(const char *pattern)
+  {
+    int error = 0;
+
+    mysql_rwlock_wrlock(&m_rwlock);
+
+    m_pattern.clear();
+    if (m_compiled)
+    {
+      my_regfree(&m_expr);
+      m_compiled = false;
+    }
+
+    if (pattern)
+    {
+      m_pattern.assign(pattern);
+    }
+
+    if (!m_pattern.empty())
+    {
+      error = my_regcomp(&m_expr,
+                         m_pattern.c_str(),
+                         MY_REG_EXTENDED | MY_REG_NOSUB | MY_REG_ICASE,
+                         &my_charset_utf8_general_ci);
```

Comment:
Looks like these rpl_do and rpl_ignore options are not treated to any charset, they are just placed into storage and later on when lookup/comparison time comes they use table_alias_charset just as you mentioned.